### DISCUSSION TITLE
[Question] Trigger on releasing a keyboard key

### DISCUSSION THREAD

**Discussion by vpoguru** [GrandOrgue/grandorgue#838] [Help]
Is it possible to trigger an audio file on disengaging a keyboard key?

I am currently porting a Hauptwerk sample set to GrandOrgue. The keyboard key attack and release noises are two separate audio files.

I implemented the engaging key action noise as follows:

    [Stop999]
    Name=Tracker Noise Keyboard
    DefaultToEngaged=Y
    WindchestGroup=999
    Percussive=Y
    NumberOfLogicalPipes=061
    NumberOfAccessiblePipes=061
    FirstAccessiblePipeLogicalPipeNumber=001
    FirstAccessiblePipeLogicalKeyNumber=001
    Pipe001=./actionNoises/keyb_attack/036-C.wav
    Pipe002=./actionNoises/keyb_attack/037-C#.wav
    ...

I tried to implement the release noise with _Pipe999Release999_, but this did not work, even if I change _Percussive_ to _N_. Does the attack sample always need a loop if not percussive? Or, is the release sample never played, when _Percussive_ is _Y_?

**Comment by vpoguru** [comment]
I have keyboard attack and release noises as two separate audio files with no loops and CUE points.

I implemented the keyboard attack noise as follows:

    [Stop998]
    Name=Tracker Noise Keyboard Attack
    DefaultToEngaged=Y
    WindchestGroup=999
    Percussive=Y
    AcceptsRetuning=N
    NumberOfLogicalPipes=061
    NumberOfAccessiblePipes=061
    FirstAccessiblePipeLogicalPipeNumber=001
    FirstAccessiblePipeLogicalKeyNumber=001
    Pipe001=./actionNoises/keyb_attack/036-C.wav
    ....

This works!

I tried to implement the release noises as follows:

    [Stop999]
    Name=Tracker Noise Keyboard Release
    DefaultToEngaged=Y
    WindchestGroup=999
    Percussive=N
    AcceptsRetuning=N
    NumberOfLogicalPipes=061
    NumberOfAccessiblePipes=061
    FirstAccessiblePipeLogicalPipeNumber=001
    FirstAccessiblePipeLogicalKeyNumber=001
    Pipe001=./actionNoises/keyb_release/036-C.wav
    Pipe001LoopCount=1
    Pipe001Loop001Start=0
    Pipe001Loop001End=1
    Pipe001CuePoint=1
    ...

This should create a very short loop at the very beginning of the wav file (frame 0 to 1) and sets the CUE point to frame 1.

This does not work. What am I missing?

**Reply by larspalo** [reply]
Loops must be in the attacks of a pipe, not in a release file!

**Reply by larspalo** [reply]
BTW, you already had the answer to this in https://github.com/GrandOrgue/grandorgue/discussions/841#discussioncomment-1641385.

**Reply by vpoguru** [reply]
> Loops must be in the attacks of a pipe, not in a release file!

I tried to use frame 0 to 1 of the release sample as attack.

**Reply by vpoguru** [reply]
> BTW, you already had the answer to this in [#841 (comment)](https://github.com/GrandOrgue/grandorgue/discussions/841#discussioncomment-1641385).

This technique has the problem that for very short notes the attack part of the sample will be truncated. This is correct for pipe sounds, but not for noises. By implementing two individual samples for attack and release, I hoped to get around this problem.

**Reply by larspalo** [reply]
> By implementing two individual samples for attack and release, I hoped to get around this problem.

If you do it right as I already told you, you'll use separate attack and release samples.

**Reply by vpoguru** [reply]
> If you do it right as I already told you, you'll use separate attack and release samples.

I used your recommendation for the implementation of the stop action noises. This worked with charm! But so far I have not been able to use the same technique for tracker noises.

**Reply by larspalo** [reply]
You can test defining the stop with the release like this:

Pipe001=DUMMY
Pipe001LoadRelease=N
Pipe001ReleaseCount=1
Pipe001Release001=actionNoises/keyb_release/036-C.wav

I've not tested this myself though, it's just an idea.

**Reply by larspalo** [reply]
> But so far I have not been able to use the same technique for tracker noises.

I did my version of Kdousov in exactly the same manner with NoisePipes output. Never thought about it twice, it just works.

Anyway, for really short staccato playing IRL I'd expect the reverb tail of the attack noise to have died before the release tail anyway and the actual attack become muffled by the release noise so quickly that you'd be hard pressed to hear any difference.

**Reply by vpoguru** [reply]
> You can test defining the stop with the release like this:
> 
> Pipe001=DUMMY Pipe001LoadRelease=N Pipe001ReleaseCount=1 Pipe001Release001=actionNoises/keyb_release/036-C.wav
> 
> I've not tested this myself though, it's just an idea.

Is it possible that the DUMMY keyword currently does not work? At least I couldn't get it working. Linking instead an empty wav file with a loop and a CUE marker did the job. Many thanks for your support!

**Reply by larspalo** [reply]
It seems that specifying DUMMY for a pipe disables any more attacks/releases and makes it useless for this purpose. I put together a demo displaying the different methods discussed here so that one can compare them.

[ActionNoise.zip](https://github.com/GrandOrgue/grandorgue/files/7613981/ActionNoise.zip)

- Stop up/left is my normal method with noisepipes.
- Stop up/right is just a percussive attack standalone
- Stop down/left is your use release file as attack with looping first samples and loading same file again as release
- Stop down/right is own dummy wav file, just a more convenient way of doing the above...

Compare stop up/left alone with the comination of stop up/right and stop down/left. Any noticeable difference?

**Comment by midimusic** [comment]
As I understand it, releases in GO do not need a loop but look at the Friesach sampleset for an example of release sample files and ODF. Personally I think this design is "over the top". One release is ample. But as far as I can tell it does work, However  GO does not handle releases well - they should not be routed via the fake reverb generator. Also in theory there should be a cross fade between the released note and the start of the release sound.

csw900

**Reply by larspalo** [reply]
> they should not be routed via the fake reverb generator

Please, show me where that is in the code. AFAIK, there's no reverb generator in GO, fake or real...

**Reply by larspalo** [reply]
The only reverb option from within GO is the usage of external impulse response files and that has nothing in particular to do with the releases.

**Reply by larspalo** [reply]
> Also in theory there should be a cross fade between the released note and the start of the release sound

Also, FYI there is a crossfade happening in GO when transitioning from the sustainsection to the release. Otherwise you'd know it...

**Comment by midimusic** [comment]
Lars

What is that item in the toolbar which sets reverb time if it is not a fake reverb generator? I do not know the GO code well enough to be able to point out where it is located. Even if there is a crossfade it is masked by the fake reverb!!!! Also short releases would be masked as well unless the reverb is set to its lowest durations.

csw900

**Reply by oleg68** [reply]
> What is that item in the toolbar which sets reverb time if it is not a fake reverb generator?

This setting limits the length of the existing release tail in the samples. But it does not generate a release tail longer than the existing one.

**Reply by larspalo** [reply]
@midimusic , trust me, there's no "fake reverb" in GO. You're just gravely mistaken about how that feature works. If you have it set on Max, there's no change done to the recorded samples and if you select a value it's the allowed max duration of a release tail. It's not how much it's shortened - it's the allowed lenght. This is a feature mostly used to remove excess reverb tail from the recorded samples when the user doesn't want to have a lot of recorded reverb (for instance if you play in a church with lots of natural reverb).

The crossfade between sustainsection and release in GO is extremely effective. You can even put releases from different pipes with other pitches and the releases will still "work" even if it sounds very strange indeed. I've tested that myself, and I've also read the code that handles the sound in GO.

**Comment by midimusic** [comment]
Hi Oleg and Lars

I do not think either of you are correct. Many organ sample sets do not have any release waveforms at all. Please explain how this fake reverb still produces adjustable reverb!! What happens if the samples are physically chopped off at the end of the loop??

Lars, you told me yourself that the demo organ samples were loaded only up to the end of the loop, how do these then produce up to 3 seconds of reverb?

Also if you have a sampleset which uses real release samples with durations of (up to say) 1 second how is it that you can still set the fake reverb to 3 seconds and get a 3 second reverb?

I do agree that the faked reverb is derived from the real sustained sound of the pipes, but otherwise its extension to 3 seconds has nothing to do with anything in the real samples.

csw900

**Reply by larspalo** [reply]
@midimusic I really don't know if it's any idea to try to explain how it works when you already "know" how... Which you actually don't, frankly!

A non percussive sample in GO consists of (at least) an attack that also have the sustainsection (loop), and hopefully (at least) one release. This release can be in the same audio file as the attack/sustain or in a separate file, and there can of course also be many different releases (and attacks)... For GO, when the sampleset is loaded, it doesn't matter if the sounds of a pipe is in one file or in many - it will be loaded correctly anyway and setup for usage according to the sampleset producers and users specifications. For RAM memory savings - GO will only load the attack from the file up to the last loop used. But of course, the releases will be loaded (and used) according to specification too! In playing, when it's time to transition to the release from the sustain, a suitable release will be selected if many are available, and there will be a kind of alignment of sustain with release and then a crossfade in between. Clear enough?

The reverb tail shortening in the toolbar can ONLY shorten a release. It CAN be set to a value longer than an actual sampled release BUT it WILL NOT elongate the release in ANY WAY. The method of shortening the release is basically a fade out. If you don't have any specific reason for setting up the reverb tail shortening in another way, it should be set to Max, to enjoy the actual recorded release of a sample unadulterated.

Why can the releases of the demo sampleset have release tails up the 3 seconds? Because they have recorded release tails with approximately that length. I can be, and have been, wrong about many things in life, but I've recorded and processed the samples for the demo sampleset myself, so this fact I do know!

**Comment by midimusic** [comment]
Hi Lars

You are quite right I do not know the details of how the reverb works in GO. This is how I am learning.

However there are two distinct cases: (You have explained only the second)

1. The samples do not have ANY recorded release waveform e.g. Barton + many others. Here it is clear that the reverb MUST be artificial (fake) because you can still get up to 3 seconds of reverb by adjusting the toolbar control. I am pretty confident the artificial release is produced by continuing to sound the loop while passing the sound through the toolbar fader. There is no crossfade because it is not necessary.

2. The samples DO have a recorded releases in which case your explanation above suffices until proven otherwise.

Many thanks for your time and effort.

csw900

**Reply by larspalo** [reply]
@midimusic That's perfectly ok.

I actually tried to explain both cases - which in reality is one and the same.

The number 1 case doesn't exist in the way you think it does. Yes, the **attack(sustain)** of a pipe can have it's audio cut after the loop, and in an audio editor you wouldn't see any release in that file (because it's not there). This kind of operation is usually done to minimize the size of the sampleset on disk and doesn't impact sound as the part of the sample after last loop wouldn't be used anyway (and optimized away by GO to reduce RAM usage). However, it's perfectly possible that the **release** is in a separate file and loaded together with the attack/sustain to form a complete **pipe**.

I cannot really comment on your statement about the Barton sampleset, all the versions I have of it have complete samples that in one file contain attack/sustain/release. Many other samplesets though use the method described above and you'll just have to look for the release separately (or read the odf to see how it's done).

Except using external IR files for adding reverb, there's no built in reverb in GO to add.

**Reply by larspalo** [reply]
Just for fun, I've put together a very small sampleset demonstrating how the different releases can be combined with a truncated attack.

[FunRelease.zip](https://github.com/GrandOrgue/grandorgue/files/7617499/FunRelease.zip)

@nickappleton This is a bit interesting on an academic level. It shows how much of the release transient actually is lost in the alignment and crossfade phase. To compensate for this, it's possible that the sampleset producer would get better release sounds in GO if the release cue was placed somewhat earlier than one normally would expect. This needs more research...

**Reply by nickappleton** [reply]
This stuff is incredibly difficult to get correct.

In my own organ sampler project, the first time an organ is loaded, the each release sample is correlated with each of the attack-sustain samples. The sampler itself computes which release should be used based on when the release should be stopped (as opposed to the sample-set creator specifying "use release X if key released before Y ms"). I found that using both an adjustable cross-fade duration *and* a target release gain that is dependent on when the note is released has given me the best results - but it's still far from perfect.

I completely agree with your statement re. release transients being lost - particularly for low frequency stops where the release transient may be short relative to the period of the waveform. I thought maybe I could start playback before the release marker - but this is problematic as many individual release samples do not contain audio before the marker - this also unfortunately impacts the ability to do good analysis in the sampler.

**Reply by larspalo** [reply]
> I thought maybe I could start playback before the release marker - but this is problematic as many individual release samples do not contain audio before the marker - this also unfortunately impacts the ability to do good analysis in the sampler.

That's why I thought this could be an interesting subject for sampleset producers cutting the releases and/or placing the cue markers. Perhaps if they are placed somewhat earlier than the actual release point the alignment and crossfade won't remove so much of the transients. And if this strategy indeed works, can we come up with a recommendation as to how much earlier a release sample should be cut, or how much earlier should a release cue be placed? Of course if too much note is left the release will feel sluggish, but there should be a sweet spot that could be targeted. 

From reading GO code it seems that the release crossfade length currently is set at 184 ms. When I get some time I'll test moving back release cue (or cut the release sample) sample_rate * 0.184 samples (frames) on a rank and we can see what happens.

**Reply by larspalo** [reply]
Moving back the release cue does indeed bring out more release transients, but at the expense of a "sticky" feel of the release. Most likely it's very hard to find a fixed value that always would produce the best result, instead it seems to me that the ear must be the final judge on where to put the release cue...

After a quick glance at the GO code it seems that the release crossfade length is scaled in between 184 and 6 ms, depending on MIDI key - unless it's specified in the odf - in which case the specified values will be used instead.

**Reply by nickappleton** [reply]
Yes, you definitely don't want to simply start playback a fixed time before the release marker. But if we could say "anywhere up to a single period of the waveform before the marker" possibly with some upper bound for very low frequency samples, that would be quite good.

As for scaling the cross-fade length: in an ideal world, the crossfade would almost always be close to zero. But we need something to avoid glitches. The crossfade length should be driven by some factor which is related to the "quality" of the release at the point that it is starting playback. In my prototype sampler, most release crossfades are on the order of 128 samples (2.7 ms), but they can vary anywhere up to 170 ms based on how confident the sampler is that the release will line up correctly.

This is really good for preserving transients - but it's quite difficult to tune.

**Comment by midimusic** [comment]
Hi Lars

Thanks for the FunRelease organ. This really does show that GO can deal correctly with releases when they are in separate files. It also shows that you have (probably!!) edited the release samples manually in order to produce a fade in. It would be interesting to check your ODF with some samples where the releases are at full amplitude, as they would normally be without editing.

However it does NOT show that GO works correctly when the releases are simply recorded in the same files as the attack/sustain, i.e attack sustain and release all following each other without any joins. i.e. the same as the GO demo organ samples.

Also it does NOT show that GO works correctly with samples containing NO release waveforms at all which in my experience seems to be  common.

It would be a very good idea if you would put together a short demo files to prove that the other two cases also work correctly.

csw900

**Reply by larspalo** [reply]
> However it does NOT show that GO works correctly when the releases are simply recorded in the same files as the attack/sustain, i.e attack sustain and release all following each other without any joins. i.e. the same as the GO demo organ samples.

For this you have plenty of samples from Burea church, Burea Funeral, Kalvträsk. (even Piteå MHS but they are wavpacked) Most of the older samplesets had samples like that. Then the samples must have loop points and a cue point for the release. This is so common that it shouldn't need any special demo.

> Also it does NOT show that GO works correctly with samples containing NO release waveforms at all which in my experience seems to be common.

If there's no release for a pipe GO **will** issue a warning in the log window, but the pipe will be playable and sound abruptly cut when released. I'm not aware of **any** GO sampleset that doesn't have release samples (in one file or many) for non percussive samples intentionally. Feel free to educate me about one that's freely available. Otherwise, I'll keep saying that this is still just a delusion.

I attach an additional odf to the FunRelease organ that just specify a third pipe that has no release. GO will warn about it (at least in strict mode) but play it back just the same. You might notice the very short fade out that would have been the crossfade into the release...
[MoreFunWithReleases.zip](https://github.com/GrandOrgue/grandorgue/files/7619018/MoreFunWithReleases.zip)

**Comment by midimusic** [comment]
Hi Lars

Your MoreFun organ does do as you say (there is no warning however) but it does not prove anything except that you know how to make a note play without any release sound.

I have considerable doubts whether the releases in the samples of any of the organs you mentioned above actually play at all. The artificial (fake) reverb which I believe GO can use would sound sufficiently realistic to be audibly just as good or better than real releases.
This may be coloured by my considerable experience with .sf2 sound fonts, the vast majority of which do not use real releases at all but can sound just as good as GO samplesets.

You have kept very quiet about the GO demo organ and this leads me to believe that the real releases in the samples are never played at all. This is what I will continue to believe until proved otherwise. It is easy to prove just by editing one of the samples to delete the release and replace it with some completely different sound and then check what happens when that note is played. When I get some spare time I will do this (if somebody else does not do it first). I have never found any release cue points in any of the GO samples I have looked at in an editor (this may be because I was not looking for them at the time). Can you suggest an editor which DOES show and is able to edit these cue points?

csw900

**Reply by larspalo** [reply]
Did you enable Strict odf checking?

Otherwise, what's your problem? The GO releases work exactly as they are supposed to do. Not without some issues and possible areas of improvement, but GO just know how to play the samples it's fed, nothing else.

I've already told about the GO demo organ, the samples come mostly from the extended part of Kalvträsk sampleset. I've recorded and edited them with some additional help from someone I don't remember right off the bat. Those samples are in one file for each pipe with loop markers and a cue marker. And yes, they are wavpacked for reducing size as much as possible.

I've programmed [LoopAuditioneer](https://sourceforge.net/projects/loopauditioneer/) for handling loops and cues in samples. This you really should know from previous posts in GO history...

**Comment by midimusic** [comment]
Hi Lars,

My problem was that there was originally no proof that anything you were saying was accurate. You then produced proof that at least some of it was correct when you published the FunRelease file. But then refused to provide proof that the single attack/sustain/release file option worked as you stated (which would have been very easy to do).

However I have now proven, by editing my GO demo samples, that the demo does indeed contain a Marker cue for each release and that the organ does play the releases in the sample files as expected. I have also proved that you do not really need any release waveforms in the sample files and reverb still plays correctly (in the artificial mode) provided the file contains a release marker cue several seconds from the end. I have also noticed that the release markers are placed well before any change of waveform indicates the real release time. This would potentially give some time for the cross fade without losing any significant amount of the release waveform. However the cross fades are of an unpublished duration (but could easily be measured).

In doing this I have now also proved that my Adobe Audition audio editor does indeed display and allow editing of these release markers, a feature I have never noticed before in spite of using the editor for well over 10 years.

Thanks Lars for your help in educating me in some of the finer points of GO.

csw900
